Scripts/Runtime/Util/Systems/WorldCache.cs

```diff
@@ -0,0 +1,187 @@
+            {
+                return;
+            }
+            m_LastRebuildCheckFrameCount = currentFrameCount;
+            
+            
+            //If we've never been built before
+            if (Version == INITIAL_VERSION)
+            {
+                Rebuild();
+                return;
+            }
+            
+            //If any of our groups have a different system count from the last time we built
+            //Someone added a new system or removed a system
+            //We can't tell if a new group was added or removed though
+            foreach (SystemGroupCache groupSystemCache in m_OrderedSystemGroupCaches)
+            {
+                if (groupSystemCache.SystemGroup.Systems.Count == groupSystemCache.CachedGroupSystemsCount)
+                {
+                    continue;
+                }
+
+                Rebuild();
+                return;
+            }
+        }
+        
+        private void Rebuild()
+        {
+            Version++;
+            m_OrderedSystemCaches.Clear();
+            m_OrderedSystemGroupCaches.Clear();
+
+            PlayerLoopSystem rootPlayerLoopSystem = PlayerLoop.GetCurrentPlayerLoop();
+            ParsePlayerLoopSystem(ref rootPlayerLoopSystem);
+        }
+
+        private void ParsePlayerLoopSystem(ref PlayerLoopSystem playerLoopSystem)
+        {
+            if (playerLoopSystem.updateDelegate != null
+             && COMPONENT_SYSTEM_GROUP_TYPE.IsAssignableFrom(playerLoopSystem.type)
+             && PlayerLoopUtil.IsPlayerLoopSystemPartOfWorld(ref playerLoopSystem, World))
+            {
+                ComponentSystemGroup group = PlayerLoopUtil.GetSystemGroupFromPlayerLoopSystemNoChecks(ref playerLoopSystem);
+                ParseSystemGroup(null, group);
+            }
+            
+            //Phases won't have subsystems
+            if (playerLoopSystem.subSystemList == null)
+            {
+                return;
+            }
+            
+            //Loop through all children
+            for (int i = 0; i < playerLoopSystem.subSystemList.Length; ++i)
+            {
+                ParsePlayerLoopSystem(ref playerLoopSystem.subSystemList[i]);
+            }
+        }
+
+        private SystemCache GetOrCreateSystemCache(SystemGroupCache groupCache, ComponentSystemBase system)
+        {
+            //See if we've ever created this SystemCache before and create if not
+            if (!m_SystemCacheLookup.TryGetValue(system, out SystemCache systemCache))
+            {
+                if (system is ComponentSystemGroup group)
+                {
+                    systemCache = new SystemGroupCache(groupCache, group);
+                }
+                else
+                {
+                    systemCache = new SystemCache(groupCache, system);
+                }
+                m_SystemCacheLookup.Add(system, systemCache);
+            }
+            m_OrderedSystemCaches.Add(systemCache);
```

Comment:
Ok i see what you're saying. 

Today it is impossible for there to be a duplicate but I could see someone unwittingly use the `GetOrCreateSystemCache` function and cause a duplicate to happen. 

I've moved the adding to the `m_OrderedSystemCaches` out to where I need them to be as part of the crawl in a rebuild.

Reply:
Brought up on call but for posterity.
Technically, it's possible to add a system to multiple `ComponentSystemGroup`s or manually add the system to the player loop in multiple locations.

Reply:
Ok fixed this up with the knowledge that Systems can be part of multiple groups. 

In that case yes we will get duplicate entries to the `m_OrderedSystemCaches` but that's ok because it's going to execute more than once.